---
title: "CITN_scRNA_preprocessing"
author: "Adam Pelletier"
date: "07/03/2022"
output: html_document
---

```{r setup, include=FALSE}

suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(Seurat))
suppressPackageStartupMessages(library(Matrix))
suppressPackageStartupMessages(library(sctransform))
#suppressPackageStartupMessages(library(hdf5r))
suppressPackageStartupMessages(library(GSVA))
suppressPackageStartupMessages(library(ggsignif))
suppressPackageStartupMessages(library(patchwork))
suppressPackageStartupMessages(library(SingleCellExperiment))
suppressPackageStartupMessages(library(MAST))
suppressPackageStartupMessages(library(GSEABase))
suppressPackageStartupMessages(library(readxl))
suppressPackageStartupMessages(library(DoubletFinder))
knitr::opts_chunk$set(echo = TRUE)
```

# CITN 12 single-cell RNA-seq analysis

### Goals
1. Preprocess and clean the data
2. Infer cell identity at a granular level : including T memory subsets
3. Determine subset frequencies that differ between the 24 hour and baseline timepoints
4. Perform differential expression on a per subset basis between the baseline and 24hr timepoint for the modules generated by Aarthi on the bulk-RNA seq to determine which subset(s) drive that enrichment.
5. Perform GSVA for visuals of those differences.
6. Identify subsets responsible for the production of the cytokines identified in bulk : IFNg, IL-4, etc. 


## Setup directories
```{r dirSetup}

setup_dir <- function(dir_string) { 
  ### Verifies creates a directory if it doesn't exist. Returns the dir_path
  if (dir.exists(dir_string)){
    dir_set <- dir_string
  } else {
    dir.create(dir_string) 
    dir_set <- dir_string
  }
  return(dir_set)
}

inputDir <- setup_dir("input")
outputDir <- setup_dir("output")
dataDir <- setup_dir(file.path(outputDir,"data"))
figureDir <- setup_dir(file.path(outputDir,"figures"))

```



## Load data


### Load raw H5 count matrix 

FASTQ files were generated by the Emory Yerkes Core, then shared to the Sekaly lab on AWS S3. 
Felipe Ten Caten shared the data to Saswat Kumar, who used the 10X Cloud service to perform preprocessing.
Felipe finished the aggregation step.  Data was shared by FTC for downstream analysis


```{r read_aggr_csv}
aggr_df <- readxl::read_excel(path = "aggregation_GEM_well_ID_appended.xlsx") %>%
              mutate(SampleNumber = gsub(".*-", "", gsub("_SI-TT-.*", "", sample_id))) %>%
              mutate(GEM_well_ID = as.character(GEM_well_ID)) 


aggr_df
```

#### Load sample keys 
```{r read sample_keys}



sample_key_df <- read_excel("input/Sample Key-2.xlsx", sheet = "sampleTimepoint" ) %>%
                  dplyr::rename(PID = "Sample ID") %>%
                  gather(Timepoint, SampleNumber, -PID) %>%
                  mutate(SampleNumber = as.character(SampleNumber))
                  
                  
      

sample_key_df
write_csv(sample_key_df, "output/sample_key_df.csv")

```


```{r read metadata}


citn_meta <- read_excel("input/CITN_all_expcept_RNAseq_Adam.xlsx") %>%
              dplyr::filter(!is.na(FlowSampleNumber)) %>%
              dplyr::rename(Days_to_EOT = "Days (EOT-C01D01)",
                            SampleNumber = "FlowSampleNumber",
                            EOT_HIV_DNA = "AVG DNA HIV/10^6 CCR5 (EOT-C01D01)") %>%
              mutate(Tumour_progression = Response) %>%
                                  mutate(Tumour_progression = ifelse(Response %in% c("Complete Response", 
                                                                                     "Stable Disease"),
                                                                     "Stable_or_Complete",
                                                                     Tumour_progression)) %>%
                                  mutate(Tumour_progression = ifelse(Tumour_progression %in% 
                                                                       c("Disease Progression",
                                                                         "Stable_or_Complete"),
                                                                         Tumour_progression, "ND")) %>%
              mutate(SampleNumber = as.character(SampleNumber)) %>%
              dplyr::select(SampleNumber, PlasmaVL, RNA_18S, DNA_CCR5, Response, Days_to_EOT, CancerType, AIDSvsNONAIDS, Cohort, CD4clinical, EOT_HIV_DNA ) %>%
              unique()
              

citn_meta

write_csv(citn_meta, file = "output/citn_metadata_vivian.csv")
```



```{r read 10x matrix}

raw <- CreateSeuratObject(Read10X_h5("filtered_feature_bc_matrix.h5"))

raw[["GEM_well_ID"]] <- gsub(".*-", "", colnames(raw))

metadata_citn_seurat <- raw@meta.data %>%
                  rownames_to_column("cell") %>%
                  dplyr::select(cell, GEM_well_ID) %>%
                  left_join(., aggr_df %>% dplyr::select(GEM_well_ID, SampleNumber), by = "GEM_well_ID") %>%
                  left_join(., sample_key_df, by = "SampleNumber") %>%
                  left_join(., citn_meta, by = "SampleNumber") %>%
                  dplyr::select(-GEM_well_ID) %>%
                  column_to_rownames("cell")

raw <- AddMetaData(raw, metadata = metadata_citn_seurat)


head(raw@meta.data)
```

### Quality Control

#### Removing low quality cells 

```{r QC RNA}

seurat <- raw
rm(raw)
DefaultAssay(seurat) <- "RNA"
seurat[["percent.mt"]] <- PercentageFeatureSet(seurat, pattern = "^MT-")
seurat[["percent.ribo"]] <- PercentageFeatureSet(seurat, pattern = "^RP[SL]")


qc_metric_plots <- sapply(c("nCount_RNA", "nFeature_RNA","percent.mt", "percent.ribo"), simplify = F, USE.NAMES = T, function(x){
    p <- VlnPlot(seurat, features = x, ncol = 1,
          log = F, pt.size = 0, group.by = "SampleNumber" ) + NoLegend()
  
})

pdf("figures/QC_metrics_ViolinPlot.pdf", width = 12)
qc_metric_plots
dev.off()

print(qc_metric_plots)

### A zoom on the lower range of NRA counts shows Sample 36 is of poor quality 
## Suggest to remove it. 
qc_metric_plots$nCount_RNA + ylim(c(0, 5000)) + ggtitle("nCount_RNA lower end zoom")
qc_metric_plots$nFeature_RNA + ylim(c(0, 3000)) + ggtitle("nFeature_RNA lower end zoom")
qc_metric_plots$percent.mt + ylim(c(0, 25)) + ggtitle("percent.mt lower end zoom")

```
*Observations*:
1. A zoom on the lower range of NRA counts shows Sample 36 is of poor quality, and 13 has a very large number of cells with low number of features 
Suggest to remove them. 

2. Number of features is a little low across the board, but we can work with that. 
Subsettting on:
1. nCount_RNA < 20000 counts
2. nCount_RNA > 1000
3. nFeatures > 200
4. percent.mt < 10

```{r subset on high_quality cells}


seurat <- subset(
  x = seurat,
  subset = nCount_RNA < 20000 &
    nCount_RNA > 1000 &
    nFeature_RNA > 200 &
    percent.mt < 10 
)

seurat_unpaired_samples <- seurat@meta.data %>%
                            dplyr::filter(!SampleNumber %in%  c("36", "34")) %>%  #### remove problematic samples
                            group_by(PID) %>%
                            summarize(n_timepoints = n_distinct(Timepoint)) %>% ###determine which Patients have only one timepoint, making paired analyses impossible. 
                            dplyr::filter(n_timepoints > 1)
  
seurat <- subset(x = seurat, subset = PID %in% seurat_unpaired_samples$PID)

saveRDS(seurat, file = "data/seurat_raw.RDS")

```
*Observations*:
1. A zoom on the lower range of NRA counts shows Sample 36 is of poor quality 
Suggest to remove it. 

Subsettting on:
1. nCount_RNA < 20000 counts
2. nCount_RNA > 1000
3. nFeatures > 200
4. percent.mt < 10



#### Removing Doublets
Using DoubletFinder for finding doublets
```{r doublet_removal, message=FALSE}
seurat <- readRDS("/mnt/efs/adam_citn/seurat_raw.RDS")


seurat <- FindVariableFeatures(seurat, verbose = F)
seurat <- ScaleData(seurat, verbose = F)
seurat <- NormalizeData(seurat,  verbose = F)
seurat <- RunPCA(seurat, verbose = F, npcs = 20)

ElbowPlot(seurat)
seurat <- RunUMAP(seurat, dims = 1:15, verbose = F)
seurat <- FindNeighbors(seurat, dims = 1:15)
seurat <- FindClusters(seurat, resolution = 0.5)


split_seurat <- SplitObject(seurat, split.by = "SampleNumber")

for(i in names(split_seurat)){
  tmp_seurat <- split_seurat[[i]]
  sweep.res <- paramSweep_v3(tmp_seurat, PCs = 1:15, sct = F) 
  sweep.stats <- summarizeSweep(sweep.res, GT = FALSE) 
  bcmvn <- find.pK(sweep.stats) 
  homotypic.prop <- modelHomotypic(tmp_seurat@meta.data$seurat_clusters) 
  print(homotypic.prop)
  nExp <- round(ncol(tmp_seurat) * homotypic.prop)  
  nExp.adj <- round(nExp*(1-homotypic.prop))
  tmp_seurat <- doubletFinder_v3(tmp_seurat, PCs = 1:15, pN = 0.25, pK = 0.09, nExp = nExp, 
                                 reuse.pANN = FALSE, sct = FALSE)
  tmp_seurat <- doubletFinder_v3(tmp_seurat, PCs = 1:15, pN = 0.25, pK = 0.09, nExp = nExp.adj, 
                             reuse.pANN = paste("pANN_0.25_0.09_", nExp, sep =""), sct = FALSE)
  print(nExp)
  print(nExp.adj)
  pANN_select <- paste("pANN_0.25_0.09_", nExp, sep ="")
  df_high <- paste("DF.classifications_0.25_0.09_", nExp.adj,sep ="")
  df_low <- paste("DF.classifications_0.25_0.09_", nExp, sep ="")
  doublet_df <- tmp_seurat@meta.data %>%
              rownames_to_column("cell") %>%
              dplyr::rename(doublet_classification_low_confidence = df_low,
                            doublet_classification_high_confidence = df_high) %>%
              dplyr::select(cell, doublet_classification_low_confidence, doublet_classification_high_confidence ) %>%
              mutate(doubletFinder_classification = doublet_classification_low_confidence) %>%
              mutate(doubletFinder_classification = ifelse(doublet_classification_high_confidence == "Doublet",
                                                 "Doublet - High Confidence", 
                                                 ifelse(doublet_classification_high_confidence == "Singlet"
                                                        & doublet_classification_low_confidence == "Doublet",
                                                        "Doublet - Low Confidence",
                                                        doubletFinder_classification))) %>% 
                                                column_to_rownames("cell")
    tmp_seurat[[pANN_select]] <- NULL
    tmp_seurat[[df_high]] <- NULL
    tmp_seurat[[df_low]] <- NULL
    tmp_seurat <- AddMetaData(tmp_seurat, metadata = doublet_df)
    split_seurat[[i]] <- tmp_seurat
}

saveRDS(split_seurat, "data/split_seurat_postDF.RDS")



seurat_filt <- merge(split_seurat[[1]], y = unlist(split_seurat)[c(2:length(split_seurat))])

rm(split_seurat)
```


```{r DF plot}
seurat_filt  <- FindVariableFeatures(seurat_filt , verbose = F)
seurat_filt  <- ScaleData(seurat_filt , verbose = F)
seurat_filt  <- NormalizeData(seurat_filt ,  verbose = F)
seurat_filt  <- RunPCA(seurat_filt , verbose = F, npcs = 20)


seurat_filt  <- RunUMAP(seurat_filt , dims = 1:15, verbose = F)
seurat_filt <- FindNeighbors(seurat_filt , dims = 1:15)
seurat_filt  <- FindClusters(seurat_filt , resolution = 0.5)


pdf("/mnt/efs/adam_citn/figures/DoubletFinder_Dimplot.pdf")
DimPlot(seurat_filt, group.by = "doubletFinder_classification", split.by = "PID", ncol = 4)
dev.off()


seurat_filt <- subset(seurat_filt,doublet_classification_low_confidence == "Singlet" )

rm(seurat)
saveRDS(seurat_filt, file = "/mnt/efs/adam_citn/data/seurat_PBMC_cleaned.RDS")
```

### Normalization and Dimension reduction

Now that problematic cells and samples are taken out, it's time to determine if there are any batch effects to correct for. 
We first perform PCA, clustering and UMAP
```{r normalization_dimreduc}
seurat_filt <- readRDS("/mnt/efs/adam_citn/data/seurat_PBMC_cleaned.RDS")

seurat_filt <- NormalizeData(seurat_filt)
seurat_filt <- FindVariableFeatures(seurat_filt, selection.method = "vst")
seurat_filt <- ScaleData(seurat_filt)



seurat_filt <- SCTransform(seurat_filt, verbose = F)


seurat_filt <- RunPCA(seurat_filt)

ElbowPlot(seurat_filt, ndims = 30)
# 25 is a good value


seurat_filt <- RunUMAP(seurat_filt, dims = 1:25)
seurat_filt <- FindNeighbors(seurat_filt, dims = 1:25)
seurat_filt <- FindClusters(seurat_filt, resolution = 0.5)

DimPlot(seurat_filt)
```

#### Exploration vs metadata variables 

No discernable batch effect
```{r data exploration, fig.width=10, fig.height=10}
pdf("figures/batch_effect_QC.pdf")
DimPlot(seurat_filt, group.by = "Timepoint", split.by = "PID", ncol = 4)
DimPlot(subset(seurat_filt, subset = Timepoint != "C01D08"), group.by = "Timepoint", split.by = "PID", ncol = 4)
DimPlot(seurat_filt, group.by = "Timepoint", split.by = "Cohort", ncol = 3)
dev.off()

```

### Save seurat object for downstream analysis
```{r saveRDS}
saveRDS(seurat_filt, file = "data/CITN_suerat_object_preproccessed.RDS")


```

---
title: "CITN_downstream_processing"
author: "Adam Pelletier"
date: "07/03/2022"
output: html_document
---

```{r setup, include=FALSE}

suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(Seurat))
suppressPackageStartupMessages(library(Matrix))
suppressPackageStartupMessages(library(sctransform))
#suppressPackageStartupMessages(library(hdf5r))
suppressPackageStartupMessages(library(GSVA))
suppressPackageStartupMessages(library(ggsignif))
suppressPackageStartupMessages(library(patchwork))
suppressPackageStartupMessages(library(SingleCellExperiment))
suppressPackageStartupMessages(library(MAST))
suppressPackageStartupMessages(library(GSEABase))
suppressPackageStartupMessages(library(readxl))
suppressPackageStartupMessages(library(DoubletFinder))
suppressPackageStartupMessages(library(celldex))
suppressPackageStartupMessages(library(SingleR))
suppressPackageStartupMessages(library(broom))
suppressPackageStartupMessages(library(RColorBrewer))
suppressPackageStartupMessages(library(pbapply))
suppressPackageStartupMessages(library(pheatmap))
suppressPackageStartupMessages(library(Azimuth))
knitr::opts_chunk$set(echo = TRUE)
```

## Load Seurat Object
```{r load seuratobject}

seurat <- readRDS("data/CITN_suerat_object_preproccessed.RDS")

```


```{r load monaco_sigs}

monaco <- MonacoImmuneData()


labels.main <- SingleR(test = seurat@assays$RNA@data,
                ref = monaco@assays@data$logcounts,
                labels = monaco@colData$label.main)

labels.main <- as.data.frame(labels.main) %>%
                  rownames_to_column("cell") %>%
                  dplyr::select(cell, labels) %>%
                  dplyr::rename(labels.main = "labels") 
                  


labels.fine <- do.call("rbind", lapply(unique(labels.main$labels.main),function(x){
    monaco_filt <- subset(monaco, select = label.main == x)
    print(x)
    sub_seurat <- subset(seurat, cells = labels.main[labels.main$labels.main == x,]$cell)
    df <- as.data.frame(SingleR(test = sub_seurat@assays$RNA@data,
                ref = monaco_filt@assays@data$logcounts,
                labels = monaco_filt@colData$label.fine)) %>%
                rownames_to_column("cell") %>%
                  dplyr::select(cell, labels) %>%
                  dplyr::rename(labels.fine = "labels")
  
}) )

singleR_order <-      labels.main %>%
                      left_join(., labels.fine, by = "cell") %>%
                      mutate(subset_class = ifelse(grepl("Dendritic|Monocyte|neutro|basoph", labels.main, ignore.case = T), "Innate", "Adaptive")) %>%
                      dplyr::select(-cell) %>%
                      unique() %>%
                      arrange(subset_class, labels.main, labels.fine) %>%
                      rowid_to_column("rank") %>%
                      mutate(labels.fine = factor(labels.fine)) %>%
                      mutate(labels.fine = reorder(labels.fine, rank)) %>%
                      dplyr::select(labels.fine, rank)
                      #column_to_rownames("cell")

singleR_metadata <- labels.main %>%
                      left_join(., labels.fine, by = "cell") %>%
                      left_join(., singleR_order, by = "labels.fine") %>%
                      mutate(labels.fine = reorder(factor(labels.fine), rank)) %>%
                      dplyr::select(-rank) %>%
                      column_to_rownames("cell")

seurat <- AddMetaData(seurat, metadata = singleR_metadata)



seurat <- subset(seurat, subset = labels.fine != "Progenitor cells")
#seurat <- subset(seurat, subset = labels.fine != "Low-density basophils")
DimPlot(seurat, group.by = "labels.main", split.by = "Timepoint")

svg("figures/CITN_singleR_monaco.svg", width = 20)

#LabelClusters(plot = DimPlot(seurat, group.by = "labels.fine", split.by = "Timepoint") + NoLegend()  , id = "labels.fine")
DimPlot(seurat, group.by = "labels.fine", split.by = "Timepoint")
dev.off()
```
```{r azimuth}

test <- seurat

ref_azimuth <- readRDS("ref.Rds")

#ref_azimuth <- SeuratData::LoadData("pbmcref", "azimuth")
# 
# DefaultAssay(ref_azimuth) <- ""
# ref_azimuth <- FindVariableFeatures(ref_azimuth, assay = "refAssay")
# ref_azimuth <- ScaleData(ref_azimuth)
# ref_azimuth <- RunPCA(ref_azimuth, assay = "refAssay", )
# 
# seurat <- FindVariableFeatures(seurat)

pbmc.anchors <- FindTransferAnchors(reference = ref_azimuth, query = seurat,
    dims = 1:30, reference.reduction = "pca", normalization.method = "SCT")
predictions <- TransferData(anchorset = pancreas.anchors, refdata = pancreas.integrated$celltype,
    dims = 1:30)
test <- LoadData("pbmcsca")
test <-  RunAzimuth(test, reference = "pbmcref", verbose = F )

```

```{r single_markers}

single_cell_subset_markers <- c("CD4", "CD8B", "CD8A", "CD19", "FCGR3A", "LYZ", "CD1C", "IL3RA", "CCR7", "IL7R", "KLRB1", "FAS", "SELL", "CCR5",
                                "CCR2", "CD27", "IGHM", "IGHD", "SPN", "CD38", "CD24", "CR2", "FCER2", "MME", "MS4A1", 
                                "PDCD1","TIGIT" ,"CD248","IL10RA", "BMPR1A", "CD5", "TCF7", "ID2", "FOXP3", "CTLA4", "RUNX1")
pdf("figures/CITN_cell_identity_markers.pdf", width = 18)
DotPlot(seurat, features = single_cell_subset_markers, group.by = "labels.fine", col.min = 0) + RotatedAxis()
dev.off()

citn_cytokine_correlates <- c("IFNG","IL17A", "IL4", "TGFB1", "IFNB1", "IFNA2", "IL6", "IL17B") 
pdf("figures/CITN_cytokine_expression.pdf")
DotPlot(seurat, features = citn_cytokine_correlates, group.by = "labels.fine", col.min = 0) + RotatedAxis()
dev.off()

bulk_tfs <- c("STAT4", "IRF8", "IRF1", "STAT3", "STAT5A", "EOMES", "GATA3", "RELA", "STAT1", "TCF7", "TCF7L2", "CTNNB1", "EZH2", "STAT6", "EGR1" ) 
pdf("figures/CITN_TF_expression.pdf", width = 10)
DotPlot(seurat, features = bulk_tfs, group.by = "labels.fine") + RotatedAxis()
dev.off()


#saveRDS(seurat, "output/data/CITN_suerat_object_preproccessed_annot.RDS")
saveRDS(seurat, "/mnt/efs/adam_citn/CITN_seurat_object_preproccessed_annot_RNA.RDS")
```

```{r calculate_frequencies}

seurat <- readRDS("/mnt/efs/adam_citn/CITN_seurat_object_preproccessed_annot_RNA.RDS")
cell_subset_frequencies <- seurat@meta.data %>%
                            rownames_to_column("cell") %>%
                            filter(labels.main != "Progenitors") %>%
                            group_by(PID, Timepoint) %>%
                            mutate(total_n = n_distinct(cell)) %>%
                            ungroup() %>%
                            group_by(PID, Timepoint, labels.fine) %>%
                            mutate(n_subset = n_distinct(cell)) %>%
                            summarize(relative_frequency = n_subset / total_n) %>%
                            ungroup() %>%
                            unique()



cell_subset_frequencies_stats <- cell_subset_frequencies %>%
                                  spread(Timepoint, relative_frequency) %>%
                                  dplyr::filter(!is.na(C01D01) & !is.na(C01D02)) %>%
                                  #gather(Timepoint, relative_frequency, -PID, -labels.fine) %>%
                                  #dplyr::filter(Timepoint != "C01D08") %>%
                                  #dplyr::filter(!is.na(C01D01) & !is.na(C01D02)) %>%
                                  #group_by(labels.fine) %>%
                                  nest(data = -labels.fine) %>%
                                  #do(model = t.test(.$C01D01, .$C01D02, paired = T)) %>%
                                  mutate(model = map(data, ~ t.test(.x$C01D01, .x$C01D02, paired = T)),
                                         tidied = map(model, tidy)) %>%
                                  unnest(tidied) %>%
                                  dplyr::filter(p.value < 0.1)

cell_subset_frequencies_plot <- cell_subset_frequencies %>%
                              filter(PID %in% paired_PID_select$`Classical monocytes`$PID) %>%
                              dplyr::filter(labels.fine %in% cell_subset_frequencies_stats$labels.fine) %>%
                              ggplot(., aes(x = Timepoint, y = relative_frequency, color = PID)) +
                                geom_point() +
                                geom_line(aes(group = PID)) +
                                theme_classic() +
                                theme(legend.position = "none",
                                      axis.text.x = element_text(angle = 45,  hjust=1)) +
                                facet_wrap(~labels.fine, scales = "free")  


tcm_frequencies <-  seurat@meta.data %>%
                            rownames_to_column("cell") %>%
                            filter(labels.main != "Progenitors") %>%
                            dplyr::filter(labels.main %in% c("CD4+ T cells", "CD8+ T cells")) %>%
                            group_by(PID, Timepoint,  labels.main) %>%
                            mutate(parent_n = n_distinct(cell)) %>%
                            ungroup() %>%
                            group_by(PID, Timepoint, labels.fine) %>%
                            mutate(n_subset = n_distinct(cell)) %>%
                            summarize(frequency_of_parent = n_subset / parent_n) %>%
                            ungroup() %>%
                            unique()
# 
#                               
# tcm_subset_frequencies_stats <- tcm_frequencies %>%
#                                   spread(Timepoint, frequency_of_parent) %>%
#                                   dplyr::filter(!is.na(C01D01) & !is.na(C01D02)) %>%
#                                   #gather(Timepoint, relative_frequency, -PID, -labels.fine) %>%
#                                   #dplyr::filter(Timepoint != "C01D08") %>%
#                                   #dplyr::filter(!is.na(C01D01) & !is.na(C01D02)) %>%
#                                   #group_by(labels.fine) %>%
#                                   nest(data = -labels.fine) %>%
#                                   #do(model = t.test(.$C01D01, .$C01D02, paired = T)) %>%
#                                   mutate(model = map(data, ~ t.test(.x$C01D01, .x$C01D02, paired = T)),
#                                          tidied = map(model, tidy)) %>%
#                                   unnest(tidied) %>%
#                                   dplyr::filter(p.value < 0.1)
# tcm_subset_frequencies_plot <- tcm_frequencies %>%
#                               dplyr::filter(labels.fine %in% tcm_subset_frequencies_stats$labels.fine) %>%
#                               ggplot(., aes(x = Timepoint, y = frequency_of_parent, color = PID)) +
#                                 geom_point() +
#                                 geom_line(aes(group = PID)) +
#                                 theme_classic() +
#                                 theme(legend.position = "none",
#                                       axis.text.x = element_text(angle = 45,  hjust=1)) +
#                                 facet_wrap(~labels.fine, scales = "free")  
# 

t_frequencies <-  seurat@meta.data %>%
                            rownames_to_column("cell") %>%
                            filter(labels.main != "Progenitors") %>%
                            dplyr::filter(grepl("T cell", labels.main)) %>%
                            group_by(PID, Timepoint) %>%
                            mutate(parent_nt = n_distinct(cell)) %>%
                            ungroup() %>%
                            group_by(PID, Timepoint, labels.fine) %>%
                            mutate(n_subset = n_distinct(cell)) %>%
                            summarize(frequency_of_cd3 = n_subset / parent_nt) %>%
                            ungroup() %>%
                            unique()
t_subset_frequencies_stats <- t_frequencies %>%
                                  spread(Timepoint, frequency_of_cd3) %>%
                                  dplyr::filter(!is.na(C01D01) & !is.na(C01D02)) %>%
                                  #gather(Timepoint, relative_frequency, -PID, -labels.fine) %>%
                                  #dplyr::filter(Timepoint != "C01D08") %>%
                                  #dplyr::filter(!is.na(C01D01) & !is.na(C01D02)) %>%
                                  #group_by(labels.fine) %>%
                                  nest(data = -labels.fine) %>%
                                  #do(model = t.test(.$C01D01, .$C01D02, paired = T)) %>%
                                  mutate(model = map(data, ~ t.test(.x$C01D01, .x$C01D02, paired = T)),
                                         tidied = map(model, tidy)) %>%
                                  unnest(tidied) %>%
                                  dplyr::filter(p.value < 0.1)

t_subset_frequencies_plot <- t_frequencies %>%
                              dplyr::filter(labels.fine %in% t_subset_frequencies_stats$labels.fine) %>%
                              ggplot(., aes(x = Timepoint, y = frequency_of_cd3, color = PID)) +
                                geom_point() +
                                geom_line(aes(group = PID)) +
                                theme_classic() +
                                theme(legend.position = "none",
                                      axis.text.x = element_text(angle = 45,  hjust=1)) +
                                facet_wrap(~labels.fine, scales = "free")  

svg("figures/relative_frequencies_total.svg")
cell_subset_frequencies_plot
dev.off()                              



Idents(seurat) <- "labels.fine"



```



## Differential Expression


### Identify baseline and C01D02 paired samples per subset

```{r sample_ID}

paired_PID_select <- seurat@meta.data %>%
                    rownames_to_column("cell") %>%
                    dplyr::filter(Timepoint %in% c("C01D01", "C01D02")) %>%
                    group_by(PID) %>%
                    mutate(n_timepoints  = n_distinct(Timepoint)) %>%
                    ungroup() %>%
                    dplyr::filter(n_timepoints == 2) %>%
                    dplyr::select(cell, PID, Timepoint, labels.fine) %>%
                    unique() %>%
                    group_by(PID, Timepoint, labels.fine) %>%
                    summarize(n_cells = n_distinct(cell)) %>%
                    ungroup() %>%
                    spread(Timepoint, n_cells, fill = 0) %>%
                    dplyr::filter(C01D01 >= 10 & C01D02 >= 10) %>%
                    group_by(labels.fine) %>%
                    mutate(n_PID = n_distinct(PID)) %>%
                    ungroup() %>%
                    dplyr::filter(n_PID >= 3) %>%
                    mutate(labels.fine = as.character(labels.fine)) %>%
                    split(., f = .$labels.fine)
              
```



```{r mast}

topTables_MAST_sub <- sapply(names(paired_PID_select), simplify = F, USE.NAMES = T, function(x){
  sub_seurat <- subset(seurat, subset = PID %in% paired_PID_select[[x]]$PID &
                              labels.fine == x & 
                              Timepoint %in% c("C01D01", "C01D02"))
  print(x)
  if(dim(sub_seurat)[2] > 5000){   ## subsample if there are mroe than 5k cells per condition.
      cells_sub <- sample(colnames(sub_seurat), size = 5000)
      sub_seurat <- sub_seurat[,cells_sub]
  }
  #return(dim(sub_seurat))
  combo <- paste(sub_seurat$labels.fine, sub_seurat$Timepoint, sep = "__")
  #return(combo)
  names(combo) <- colnames(sub_seurat)
  #return(combo)
  sub_seurat[["combo"]] <- combo
  Idents(sub_seurat) <- "combo"
  df <- FindMarkers(sub_seurat, ident.1 = paste(x, "C01D02", sep = "__"),
                                ident.2 = paste(x, "C01D01", sep = "__"), test.use = "MAST",
                    logfc.threshold = 0,
                    min.pct = 0,
                    only.pos = F) %>%
        rownames_to_column("gene")
})

saveRDS(topTables_MAST_sub, "/mnt/efs/adam_citn/topTables_MAST_subsampled.RDS")



consolidated_MAST <- do.call("rbind", lapply(names(topTables_MAST_sub), function(x){
  df <- topTables_MAST_sub[[x]] %>%
        mutate(cell_type = x)
  
  
}))

```


```{r gsva}

modules_bulk <- read_excel("C01D01vsC01D02_Modules.xlsx") 


moduleLS <- sapply(modules_bulk$Function, simplify = F, USE.NAMES = T,  function(x){
        ledge <- modules_bulk %>%
                  dplyr::filter(Function == x) %>%
                  separate_rows(LEADING_edge, sep = ",") 
        
        return(ledge$LEADING_edge)
})

gmt_modules <- GeneSetCollection(lapply(names(moduleLS), function(x){
  ledge <- moduleLS[[x]]
  gs <- GeneSet(geneIds = ledge, setName = x)
  
  
}))

```

